### ISSUE TITLE
InterruptionHandling broken in ActionImpl

### ISSUE THREAD

**Issue by DivineThreepwood** [openbase/bco#87]
It seems like the InterruptedException triggered by an action cancelation gets lost and never reaches the action task that is currently processing the operation service state update. Thus, the entire unit controller stucks that causes serveral further issues.

* Branch: `bugfix/#87_lost_interruption_in_action_impl`
* How to reproduce: just perform `AbsenceEnergySavingAgentTest`

```
22:01:30.301 [pool-1-thread-66] INFO  o.o.b.d.c.l.u.l.LocationControllerImpl Paradise is updated from absent to present.
22:01:30.352 [Test worker] INFO  o.o.b.d.r.l.u.u.UnitStateAwaiter Woke up! Time waited 109ms
22:01:30.364 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[CANCELING] ActionImpl[action-normal-8664c4f8|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-absence_energy_saving_agent_unit_test_stairway_to_heaven-330d0966(SYSTEM)|EXECUTING]
22:01:30.364 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[CANCELED] ActionImpl[action-normal-8664c4f8|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-absence_energy_saving_agent_unit_test_stairway_to_heaven-330d0966(SYSTEM)|CANCELING]
22:01:30.365 [pool-1-thread-98] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.365 [pool-1-thread-98] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:30.366 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[INITIATING] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|SCHEDULED]
22:01:30.378 [pool-1-thread-97] INFO  o.o.b.d.c.a.ActionImpl State[SUBMISSION] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|INITIATING]
22:01:30.379 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.379 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:30.390 [pool-1-thread-97] ERROR o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] performOperationService: 314432329
22:01:30.392 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[INITIALIZED] ActionImpl[action-low-fc9e34a1|POWER_STATE_SERVICE|{"value": "ON"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|UNKNOWN]
22:01:30.393 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[ABORTING] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|SUBMISSION]
22:01:30.393 [pool-1-thread-97] ERROR o.o.b.d.c.l.u.ColorableLightController wait for lock...
22:01:30.393 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.398 [pool-1-thread-98] WARN  o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] cancel[true] and interrupt 314432329
22:01:30.498 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:30.498 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:30.498 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.602 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:30.602 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:30.602 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.707 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:30.707 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:30.707 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.812 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:30.812 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:30.812 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:30.916 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:30.916 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:30.916 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.021 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:31.021 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:31.021 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.126 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:31.126 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:31.126 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.231 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:31.231 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:31.231 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.335 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it false
22:01:31.335 [pool-1-thread-97] WARN  o.o.b.d.c.l.u.ColorableLightController stucked
22:01:31.335 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.382 [pool-1-thread-82] WARN  o.o.b.d.r.l.u.ColorableLightRemote Remote connection to Controller[/location-1/location-3/colorablelight/colorablelight-3] lost!
22:01:31.403 [pool-1-thread-98] ERROR o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] Can not cancel ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|ABORTING] it seems the execution has stuck.
22:01:31.409 [pool-1-thread-98] WARN  o.o.b.d.c.a.ActionImpl Thread[pool-1-thread-98] state[RUNNABLE]
=== Stacktrace ===
java.base@17.0.5/java.lang.Thread.dumpThreads(Native Method)
java.base@17.0.5/java.lang.Thread.getAllStackTraces(Thread.java:1662)
app//org.openbase.jul.exception.StackTracePrinter.printAllStackTraces(StackTracePrinter.java:145)
app//org.openbase.bco.dal.control.action.ActionImpl.cancelActionTask(ActionImpl.java:742)
app//org.openbase.bco.dal.control.action.ActionImpl.abort(ActionImpl.java:575)
app//org.openbase.bco.dal.lib.action.SchedulableAction.abort(SchedulableAction.java:64)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.reschedule(AbstractUnitController.java:1080)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.scheduleAction(AbstractUnitController.java:858)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.applyAction(AbstractUnitController.java:675)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.internalApplyActionAuthenticated(AbstractUnitController.java:1278)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.lambda$applyActionAuthenticated$5(AbstractUnitController.java:1226)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController$$Lambda$724/0x0000000801477c90.process(Unknown Source)
app//org.openbase.bco.authentication.lib.AuthenticatedServiceProcessor.authenticatedAction(AuthenticatedServiceProcessor.java:135)
app//org.openbase.bco.authentication.lib.AuthenticatedServiceProcessor.authenticatedAction(AuthenticatedServiceProcessor.java:90)
app//org.openbase.bco.authentication.lib.AuthenticatedServiceProcessor.authenticatedAction(AuthenticatedServiceProcessor.java:69)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.applyActionAuthenticated(AbstractUnitController.java:1244)
java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
java.base@17.0.5/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.base@17.0.5/java.lang.reflect.Method.invoke(Method.java:568)
app//kotlin.reflect.jvm.internal.calls.CallerImpl$Method.callMethod(CallerImpl.kt:97)
app//kotlin.reflect.jvm.internal.calls.CallerImpl$Method$Instance.call(CallerImpl.kt:113)
app//kotlin.reflect.jvm.internal.KCallableImpl.call(KCallableImpl.kt:108)
app//org.openbase.jul.communication.mqtt.RPCMethod.invoke(RPCMethod.kt:223)
app//org.openbase.jul.communication.mqtt.RPCServerImpl.handleRemoteCall(RPCServerImpl.kt:141)
app//org.openbase.jul.communication.mqtt.RPCServerImpl.activate$lambda-2$lambda-1(RPCServerImpl.kt:65)
app//org.openbase.jul.communication.mqtt.RPCServerImpl$$Lambda$550/0x0000000801130aa0.accept(Unknown Source)
app//com.hivemq.client.internal.mqtt.MqttAsyncClient$CallbackSubscriber.onNext(MqttAsyncClient.java:303)
app//com.hivemq.client.internal.mqtt.MqttAsyncClient$CallbackSubscriber.onNext(MqttAsyncClient.java:288)
app//com.hivemq.client.rx.FlowableWithSingle$SingleFutureSubscriber.onNext(FlowableWithSingle.java:406)
app//com.hivemq.client.internal.rx.operators.FlowableWithSingleCombine$SplitSubscriber$Default.tryOnNextActual(FlowableWithSingleCombine.java:235)
app//com.hivemq.client.internal.rx.operators.FlowableWithSingleCombine$SplitSubscriber.tryOnNext(FlowableWithSingleCombine.java:200)
app//io.reactivex.internal.operators.flowable.FlowableObserveOn$ObserveOnConditionalSubscriber.runAsync(FlowableObserveOn.java:649)
app//io.reactivex.internal.operators.flowable.FlowableObserveOn$BaseObserveOnSubscriber.run(FlowableObserveOn.java:176)
app//io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:288)
app//io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:253)
java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
java.base@17.0.5/java.lang.Thread.run(Thread.java:833)
==================
22:01:31.410 [pool-1-thread-98] WARN  o.o.b.d.c.a.ActionImpl Thread[pool-1-thread-97] state[TIMED_WAITING]
=== Stacktrace ===
java.base@17.0.5/jdk.internal.misc.Unsafe.park(Native Method)
java.base@17.0.5/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:252)
java.base@17.0.5/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:717)
java.base@17.0.5/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireNanos(AbstractQueuedSynchronizer.java:986)
java.base@17.0.5/java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.tryLock(ReentrantReadWriteLock.java:1131)
app//org.openbase.jul.extension.protobuf.BuilderSyncSetup.tryLockWrite(BuilderSyncSetup.java:177)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.applyDataUpdate(AbstractUnitController.java:1377)
app//org.openbase.bco.dal.lib.layer.unit.UnitController.applyDataUpdate(UnitController.java:79)
app//org.openbase.bco.dal.lib.layer.unit.UnitController.applyServiceState(UnitController.java:120)
app//org.openbase.bco.dal.lib.layer.unit.UnitController.applyServiceState(UnitController.java:99)
app//org.openbase.bco.dal.lib.layer.unit.UnitController.applyServiceState(UnitController.java:141)
app//org.openbase.bco.dal.lib.layer.service.mock.OperationServiceFactoryMock.update(OperationServiceFactoryMock.java:261)
app//org.openbase.bco.dal.lib.layer.service.mock.OperationServiceFactoryMock$PowerStateOperationServiceMock.setPowerState(OperationServiceFactoryMock.java:168)
java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
java.base@17.0.5/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.base@17.0.5/java.lang.reflect.Method.invoke(Method.java:568)
app//org.openbase.bco.dal.lib.layer.service.Services.invokeServiceMethod(Services.java:546)
app//org.openbase.bco.dal.lib.layer.service.Services.invokeServiceMethod(Services.java:522)
app//org.openbase.bco.dal.lib.layer.service.Services.invokeOperationServiceMethod(Services.java:569)
app//org.openbase.bco.dal.control.layer.unit.AbstractUnitController.performOperationService(AbstractUnitController.java:2007)
app//org.openbase.bco.dal.control.action.ActionImpl.lambda$execute$0(ActionImpl.java:304)
app//org.openbase.bco.dal.control.action.ActionImpl$$Lambda$672/0x00000008013265b8.call(Unknown Source)
java.base@17.0.5/java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
java.base@17.0.5/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
java.base@17.0.5/java.lang.Thread.run(Thread.java:833)
==================
22:01:31.411 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[SCHEDULED] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|ABORTING]
22:01:31.411 [pool-1-thread-98] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.411 [pool-1-thread-98] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:31.411 [pool-1-thread-98] INFO  o.o.b.d.c.a.ActionImpl State[INITIATING] ActionImpl[action-low-fc9e34a1|POWER_STATE_SERVICE|{"value": "ON"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|INITIALIZED]
22:01:31.412 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:31.412 [pool-1-thread-97] INFO  o.o.b.d.c.l.u.ColorableLightController PH Hue E27 Device Stairway is updated from off to off.
22:01:31.416 [pool-1-thread-97] ERROR o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] performOperationService end: 314432329
22:01:31.420 [pool-1-thread-84] INFO  o.o.b.d.c.a.ActionImpl State[REJECTED] ActionImpl[action-low-fc9e34a1|POWER_STATE_SERVICE|{"value": "ON"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|INITIATING]
22:01:31.421 [pool-1-thread-84] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.421 [pool-1-thread-84] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:31.463 [pool-2-thread-2] INFO  o.o.b.d.c.a.ActionImpl State[INITIATING] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|SCHEDULED]
22:01:31.469 [pool-1-thread-97] INFO  o.o.b.d.c.a.ActionImpl State[SUBMISSION] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|INITIATING]
22:01:31.470 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.470 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:31.477 [pool-1-thread-97] ERROR o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] performOperationService: 31171132
22:01:31.480 [pool-1-thread-97] ERROR o.o.b.d.c.l.u.ColorableLightController wait for lock...
22:01:31.480 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.480 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:31.480 [pool-1-thread-97] INFO  o.o.b.d.c.l.u.ColorableLightController PH Hue E27 Device Stairway is updated from off to off.
22:01:31.485 [pool-1-thread-97] ERROR o.o.b.d.c.a.ActionImpl [action-termination-economy-144d8fed - BCO changed power of PH Hue E27 Device Stairway to off.] performOperationService end: 31171132
22:01:31.485 [pool-1-thread-97] INFO  o.o.b.d.c.a.ActionImpl BCO changed power of PH Hue E27 Device Stairway to off.
22:01:31.485 [pool-1-thread-97] INFO  o.o.b.d.c.a.ActionImpl State[EXECUTING] ActionImpl[action-termination-economy-144d8fed|POWER_STATE_SERVICE|{"value": "OFF"}|colorablelight-14f2bbe8|user-bco-185a8dcd(SYSTEM)|SUBMISSION]
22:01:31.485 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup try lock
22:01:31.485 [pool-1-thread-97] WARN  o.o.j.e.p.BuilderSyncSetup got it true
22:01:32.309 [pool-1-thread-84] INFO  o.o.b.d.r.l.u.ColorableLightRemote Connection reestablished ColorableLightRemote[scope:/location-1/location-3/colorablelight/colorablelight-3]
22:01:32.381 [pool-1-thread-69] INFO  o.o.j.s.r.ProtoBufFileSynchronizedRegistry [AgentConfigRegistry]: Remove UnitConfig[Agent-1:Absence Energy Saving Agent Unit Test]...
22:01:32.381 [pool-1-thread-69] INFO  o.o.j.s.r.ProtoBufFileSynchronizedRegistry [UserConfigRegistry]: Remove UnitConfig[User-5:Absence Energy Saving Agent Unit Test@Stairway to Heaven]...
22:01:32.388 [pool-1-thread-69] INFO  o.o.j.s.r.ProtoBufFileSynchronizedRegistry [LocationConfigRegistry]: 100% consistency checks passed of LocationConfigRegistry after 1 applied modifications.
22:01:32.401 [pool-1-thread-69] INFO  o.o.j.s.r.ProtoBufFileSynchronizedRegistry [AuthorizationGroupConfigRegistry]: 100% consistency checks passed of AuthorizationGroupConfigRegistry after 1 applied modifications.
22:01:32.412 [pool-1-thread-69] INFO  o.o.j.s.r.ProtoBufFileSynchronizedRegistry [LocationConfigRegistry]: 100% consistency checks passed of LocationConfigRegistry after 2 applied modifications.
22:01:32.456 [Test worker] INFO  o.o.a.t.a.AbstractBCOAgentManagerTest Cancel all ongoing actions...
22:01:32.508 [Test worker] INFO  o.o.b.d.t.AbstractBCOTest Cancel 1 ongoing test action ...
```

**Comment by pLeminoq** [comment]
The issue was that the `actionTask` was cancelled before (when updating the action state to aborting) without the interruption flag. As the [javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#cancel%28boolean%29) of the Future states, once a future is done or cancelled the next call to cancelling will not do anything (fail).

**Comment by DivineThreepwood** [comment]
Issue fixed in #88
